soccer_ipm/soccer_ipm/soccer_ipm.py

```diff
@@ -0,0 +1,323 @@
+        self._obstacle_footpoint_out_of_image_threshold = \
+            self.get_parameter('obstacles.footpoint_out_of_image_threshold').value
+        self._goalpost_footpoint_out_of_image_threshold = \
+            self.get_parameter('goalposts.footpoint_out_of_image_threshold').value
+        camera_info_topic = self.get_parameter('camera_info.camera_info_topic').value
+        balls_in_image_topic = self.get_parameter('ball.ball_topic').value
+        goalposts_in_image_topic = self.get_parameter('goalposts.goalposts_topic').value
+        obstacles_in_image_topic = self.get_parameter('obstacles.obstacles_topic').value
+        field_boundary_in_image_topic = \
+            self.get_parameter('field_boundary.field_boundary_topic').value
+        line_mask_in_image_topic = self.get_parameter('masks.line_mask.topic').value
+        line_mask_scaling = self.get_parameter('masks.line_mask.scale').value
+
+        # Subscribe to camera info
+        self.create_subscription(CameraInfo, camera_info_topic, self.ipm.set_camera_info, 1)
+
+        self.balls_pub = self.create_publisher(
+            sv3dm.BallArray, 'balls_relative', 1)
+        self.line_mask__pub = self.create_publisher(
+            PointCloud2, 'line_mask_relative_pc', 1)
+        self.goalposts_pub = self.create_publisher(
+            sv3dm.GoalpostArray, 'goal_posts_relative', 1)
+        self.robots__pub = self.create_publisher(
+            sv3dm.RobotArray, 'robots_relative', 1)
+        self.field_boundary_pub = self.create_publisher(
+            sv3dm.FieldBoundary, 'field_boundary_relative', 1)
+
+        # Subscribe to image space data topics
+        self.create_subscription(
+            BallArray, balls_in_image_topic, self.callback_ball, 1)
+        self.create_subscription(
+            GoalpostArray, goalposts_in_image_topic, self.callback_goalposts, 1)
+        self.create_subscription(
+            RobotArray, obstacles_in_image_topic, self.callback_robots, 1)
+        self.create_subscription(
+            FieldBoundary, field_boundary_in_image_topic, self.callback_field_boundary, 1)
+        self.create_subscription(
+            Image,
+            line_mask_in_image_topic,
+            lambda msg: self.callback_masks(
+                msg,
+                self.line_mask__pub,
+                scale=line_mask_scaling), 1)
+
+    def get_field(self, time, heigh_offset=0):
+        plane = PlaneStamped()
+        plane.header.frame_id = self._base_footprint_frame
+        plane.header.stamp = time
+        plane.plane.coef[2] = 1.0  # Normal in z direction
+        plane.plane.coef[3] = -heigh_offset  # Distance above the ground
+        return plane
+
+    def callback_ball(self, msg: BallArray):
+        field = self.get_field(msg.header.stamp, self._ball_height)
+
+        balls_relative = sv3dm.BallArray()
+        balls_relative.header.stamp = msg.header.stamp
+        balls_relative.header.frame_id = self._base_footprint_frame
```

Comment:
In the case of 3d vision (through a stereo or depth camera), these would be relative to the optical frame of the camera. Would it make more sense if these are all published relative to the optical frame of the camera here too?

Reply:
You could set the frame to be exactly like the 3d camera, if you set the `base_footprint_frame` (maybe we could rename it to output frame) to the camera optical frame. But I see no practical benefits in this. Stereo cams do this because they just want to be a sensor and don't want to depend on the forward kinematics. We depend on it ether way, so why remove it and add more complexity and latency to the pipeline by routing it through another transform node which would need to be created for our messages in the first place. Having these measurements in the global frame also allows for much easier handling if e.g. the camera moved, you don't have a dedicated transformation node, and you don't want to have extremely long TF buffers to compensate for it at a later point at the subscriber side.

Reply:
Considering tracking algorithms usually run in ``base_footprint``, both for wheeled and humanoid robots, I like your suggestion of leaving this as base_footprint for (hopefully) less computation.

However, I don't think other modules (ie. ball tracking kalman filter) should assume that the frame_id is always going to be base_footprint, because then it wouldn't work for stereo camera 3d vision. The receiving end should always have code to transform messages into their desired frame of reference and not assume it's sv3dm messages are always in ``base_footprint``.